### PR TITLE
implement fuse reduce->unsqueeze + fix assumption in nop_dropout pass

### LINKED ISSUE #1564: Optimization: Fuse {Reduction[keepdim=0]->Unsqueeze} to {Reduction[keepdim=1]}

### PR DESCRIPTION
**Pull request by ArmenAg**
fixes #1564

### CODE REVIEW

**Comment by ArmenAg** [comment]
@houseroad could you take a look please

File: onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    if (cur_node_check) {
+      Node* prev_node = node->input()->node();
+      // check that the previous node a reduction operator and has defined
+      // axes/keepdims
+      bool reduction_node_check = reduction_operators.find(prev_node->kind()) !=
+              reduction_operators.end() &&
+          prev_node->hasAttribute(kaxes) && prev_node->hasAttribute(kkeepdims);
+      if (reduction_node_check) {
+        // insure that keepdims is set to false currently
+        return prev_node->i(kkeepdims) != 1 &&
```
Comment (houseroad) [inline onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h]: better to use == 0 instead of != 1

File: onnx/test/optimizer_test.py [hunk 2]
```diff
@@ -1120,6 +1120,34 @@ def test_eliminate_nop_dropout(self):  # type: () -> None
...
+                for axes2 in [[1], [1, 2]]:
+                    for keepdim in [False, True]:
+                        node = helper.make_node(reduction, ["X"], ["Y"], axes=axes1, keepdims=keepdim)
+                        node1 = helper.make_node("Unsqueeze", ["Y"], ["Z"], axes=axes2)
+                        graph = helper.make_graph(
+                            [node, node1],
+                            "test",
+                            [helper.make_tensor_value_info(
+                                "X", TensorProto.FLOAT, (5, 7, 9))],
+                            [helper.make_tensor_value_info("Z", TensorProto.FLOAT, (5, 9))])
```
Comment (houseroad) [inline onnx/test/optimizer_test.py]: output shape are not valid in some cases

File: onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h (hunk 1 above)
Comment (ArmenAg) [reply onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h]: Good call

File: onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  explicit FuseConsecutiveReduceUnsqueeze()
+      : PredicateBasedPass(
+            PassType::Fuse,
+            PassEfficiency::Complete,
+            PassOptimizationType::Compute) {}
+
+  std::string getPassName() const override {
+    return "fuse_consecutive_reduce_unsqueeze";
+  }
+  static bool axes_match(
```
[the remark this answers is not in the document]
Comment (ArmenAg) [reply onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h]: Oops. Forgot to remove this. I was using this to debug some stuff. My bad.

File: onnx/test/optimizer_test.py (hunk 2 above)
Comment (ArmenAg) [reply onnx/test/optimizer_test.py]: fixed.

**Comment by ArmenAg** [comment]
@houseroad all fixed.

File: onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
```
Comment (houseroad) [inline onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h]: why do we still keep `axes_match`?

Comment (ArmenAg) [reply onnx/optimizer/passes/fuse_consecutive_reduce_unsqueeze.h]: Sorry. My bad. Deleted it now.

File: onnx/test/optimizer_test.py [hunk 5]
```diff
@@ -1120,6 +1120,38 @@ def test_eliminate_nop_dropout(self):  # type: () -> None
...
+        for reduction in ["ReduceL1", "ReduceL2", "ReduceLogSum",
+                          "ReduceLogSumExp", "ReduceMax", "ReduceMean",
+                          "ReduceMin", "ReduceProd", "ReduceSum", "ReduceSumSquare"]:
+            for axes1 in [[1], [1, 2]]:
+                for axes2 in [[1], [1, 2]]:
+                    for keepdim in [False, True]:
+                        input_shape = (5, 7, 9)
+                        output_shape = input_shape  # type: Tuple[int, ...]
+                        if not keepdim and axes1 == axes2:
+                            output_shape = tuple([x for i, x in enumerate(output_shape) if i not in axes1])
```
Comment (houseroad) [inline onnx/test/optimizer_test.py]: output_shape seems still incorrect.
For example:
axes1 = [1]
axes2 = [1]
and keepdim = false
outshape should be (5, 1, 9)

Comment (ArmenAg) [reply onnx/test/optimizer_test.py]: fixed. I missed a ton of cases. Sorry about that.

**Comment by ArmenAg** [comment]
@houseroad Looks like some tests are broken on the PyTorch side.

File: onnx/test/optimizer_test.py [hunk 6]
```diff
@@ -1120,6 +1120,47 @@ def test_eliminate_nop_dropout(self):  # type: () -> None
...
+                                "X", TensorProto.FLOAT, input_shape)],
+                            [helper.make_tensor_value_info("Z", TensorProto.FLOAT, output_shape)])
+                        optimized_model = self._optimized(
+                            graph, ["fuse_consecutive_reduce_unsqueeze"], False)
+
+                        if keepdim or axes1 != axes2:
+                            assert optimized_model.graph == graph
+                        else:
+                            assert len(optimized_model.graph.output) == 1
+                            assert len(optimized_model.graph.node) == 1
```
Comment (houseroad) [inline onnx/test/optimizer_test.py]: Let's also check the output shape please, since we will change it in this opt pass.

Comment (houseroad) [reply onnx/test/optimizer_test.py]: Yeah, they are what I was looking for

**Pull request merged**